File: Python/5kyu/my-smallest-code-interpreter-aka-brainfk.py
"""
Title: My smallest code interpreter (aka Brainf**k)
Link: https://www.codewars.com/kata/526156943dfe7ce06200063e
Difficulty: 5 kyu

## Description

Implement a Brainfuck interpreter: tape of 8-bit wrapping cells, up to 65536 cells; `>` extends the
tape or wraps the pointer to 0; `<` at position 0 wraps to the end; `,` consumes `program_input`;
`.` appends to the output. Ignore any character other than `><+-.,[]`.
"""

from dataclasses import dataclass, field

MAX_CELLS = 65536


def _build_bracket_jumps(source: str) -> dict[int, int]:
    open_stack: list[int] = []
    bracket_jumps: dict[int, int] = {}
    for idx, ch in enumerate(source):
        if ch == "[":
            open_stack.append(idx)
        elif ch == "]":
            open_at = open_stack.pop()
            bracket_jumps[open_at] = idx
            bracket_jumps[idx] = open_at
    return bracket_jumps


def _shift_right(ptr: int, cells: list[int]) -> int:
    ptr += 1
    if ptr < len(cells):
        return ptr
    if len(cells) < MAX_CELLS:
        cells.append(0)
        return ptr
    return 0


def _shift_left(ptr: int, cells: list[int]) -> int:
    return len(cells) - 1 if ptr == 0 else ptr - 1


def _increment_cell(cells: list[int], ptr: int) -> None:
    cells[ptr] = (cells[ptr] + 1) & 255


def _decrement_cell(cells: list[int], ptr: int) -> None:
    cells[ptr] = (cells[ptr] - 1) & 255


@dataclass
class _InterpreterState:
    cells: list[int] = field(default_factory=lambda: [0])
    ptr: int = 0
    ip: int = 0
    in_pos: int = 0
    out: list[str] = field(default_factory=list)


def _apply_bracket_jump(
    state: _InterpreterState,
    bracket_jumps: dict[int, int],
    *,
    if_zero: bool,
) -> None:
    """`[` uses if_zero=True; `]` uses if_zero=False."""
    ip = state.ip
    cell_is_zero = state.cells[state.ptr] == 0
    do_jump = cell_is_zero if if_zero else not cell_is_zero
    state.ip = bracket_jumps[ip] if do_jump else ip


def _execute_step(
    source: str,
    state: _InterpreterState,
    bracket_jumps: dict[int, int],
    inp: str,
) -> None:
    match source[state.ip]:
        case ">":
            state.ptr = _shift_right(state.ptr, state.cells)
        case "<":
            state.ptr = _shift_left(state.ptr, state.cells)
        case "+":
            _increment_cell(state.cells, state.ptr)
        case "-":
            _decrement_cell(state.cells, state.ptr)
        case ".":
            state.out.append(chr(state.cells[state.ptr]))
        case ",":
            state.cells[state.ptr] = ord(inp[state.in_pos])
            state.in_pos += 1
        case "[":
            _apply_bracket_jump(state, bracket_jumps, if_zero=True)
        case "]":
            _apply_bracket_jump(state, bracket_jumps, if_zero=False)
        case _:
            pass


def brain_luck(code: str, program_input: str) -> str:
    bracket_jumps = _build_bracket_jumps(code)
    state = _InterpreterState()
    while state.ip < len(code):
        _execute_step(code, state, bracket_jumps, program_input)
        state.ip += 1
    return "".join(state.out)

File: languages/javascript/5kyu/my-smallest-code-interpreter-aka-brainfk.js
/**
 * Title: My smallest code interpreter (aka Brainf**k)
 * Link: https://www.codewars.com/kata/526156943dfe7ce06200063e
 * Difficulty: 5 kyu
 *
 * ## Description
 *
 * Implement a Brainfuck interpreter: tape of 8-bit wrapping cells, up to 65536 cells; `>` extends
 * the tape or wraps the pointer to 0; `<` at position 0 wraps to the end; `,` consumes `input`;
 * `.` appends to the output. Ignore any characters other than `><+-.,[]`.
 */

const MAX_CELLS = 65536;

function buildBracketJumps(source) {
  const openStack = [];
  const jumps = new Map();
  for (let i = 0; i < source.length; i++) {
    const ch = source[i];
    if (ch === '[') openStack.push(i);
    else if (ch === ']') {
      const openAt = openStack.pop();
      jumps.set(openAt, i);
      jumps.set(i, openAt);
    }
  }
  return jumps;
}

function shiftRight(ptr, cells) {
  ptr += 1;
  if (ptr < cells.length) return ptr;
  if (cells.length < MAX_CELLS) {
    cells.push(0);
    return ptr;
  }
  return 0;
}

function shiftLeft(ptr, cells) {
  return ptr === 0 ? cells.length - 1 : ptr - 1;
}

function nextBracketIp(s, ifZero) {
  const cellIsZero = s.cells[s.ptr] === 0;
  const doJump = ifZero ? cellIsZero : !cellIsZero;
  return doJump ? s.jumps.get(s.ip) : s.ip;
}

const tapeOps = {
  '>': (m) => {
    m.ptr = shiftRight(m.ptr, m.cells);
  },
  '<': (m) => {
    m.ptr = shiftLeft(m.ptr, m.cells);
  },
  '+': (m) => {
    m.cells[m.ptr] = (m.cells[m.ptr] + 1) & 255;
  },
  '-': (m) => {
    m.cells[m.ptr] = (m.cells[m.ptr] - 1) & 255;
  },
};

function runStep(m, code, input) {
  const ch = code[m.ip];
  const tapeOp = tapeOps[ch];
  if (tapeOp) {
    tapeOp(m);
    return;
  }
  if (ch === '.') {
    m.out += String.fromCharCode(m.cells[m.ptr]);
    return;
  }
  if (ch === ',') {
    m.cells[m.ptr] = input.charCodeAt(m.inPos++);
    return;
  }
  if (ch === '[') {
    m.ip = nextBracketIp(
      { cells: m.cells, ptr: m.ptr, ip: m.ip, jumps: m.jumps },
      true,
    );
    return;
  }
  if (ch === ']') {
    m.ip = nextBracketIp(
      { cells: m.cells, ptr: m.ptr, ip: m.ip, jumps: m.jumps },
      false,
    );
  }
}

function brainLuck(code, input) {
  const m = {
    jumps: buildBracketJumps(code),
    cells: [0],
    ptr: 0,
    ip: 0,
    inPos: 0,
    out: '',
  };
  while (m.ip < code.length) {
    runStep(m, code, input);
    m.ip += 1;
  }
  return m.out;
}
